tiledb/api/c_api_support/cpp_string/cpp_string.h

```diff
@@ -0,0 +1,156 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ */
+
+#ifndef TILEDB_CAPI_SUPPORT_CPP_STRING_H
+#define TILEDB_CAPI_SUPPORT_CPP_STRING_H
+
+#include <cstring>                   // std::memchr
+#include <string>                    // std::string_view
+#include "../argument_validation.h"  // CAPIException
+
+namespace tiledb::api {
+
+/**
+ * The default maximum length of C strings accepted by the C API
+ *
+ * This value is used as a template argument, should there ever be a need to
+ * use something other than the default length. A large class of uses are for
+ * the names of query fields; at present these do not have a defined maximum
+ * admissible length, but ought to have. This default value is large enough for
+ * most purposes.
+ */
+constexpr size_t default_max_c_string_length = 65534;
+
+/**
+ * Validate that a pointer is to a C-style string and convert it to a string
+ * view.
+ *
+ * This function is a defense against malformed inputs coming in through the C
+ * API. This function validates the following properties of the candidate input:
+ *   - The pointer is not null. (It points to something.)
+ *   - The string is null-terminated. (It has a length.)
+ *   - The string length is admissible. (It's short enough.)
+ * The admissible string length might vary depending on how it's being used.
+ *
+ * This function does not throw. It returns an empty string if there's a
+ * validation error.
+ *
+ * @section Maturity
+ *
+ * There's one failure that this function cannot withstand at present. If the
+ * pointer is to a non-terminated string at the end of a page, there will be an
+ * access violation fault that necessarily happens during the search for a
+ * terminating NUL character. It would be possible to recover from such an error
+ * by catching the resulting signal and causing this function to simply return
+ * false. This behavior, however, is not nearly as simple as it might appear.
+ *
+ * The proper requirement for such behavior is that it be able to deal with
+ * multiple simultaneous invocations of the function. As this function is
+ * defensive, it should be able to handle many simultaneous calls correctly, as
+ * an intentional attacker might generate them specifically with an eye to
+ * cause the defense to fault.
+ *
+ * @param candidate_c_string A pointer to be validated as a null-terminated
+ *   character string
+ * @return A string view corresponding to the candidate if it is valid. A null
+ *   string view (`.data() == nullptr`) otherwise.
+ */
```

Comment:
IMHO if we need validating string arguments beyond a null check, an `std::string_view(candidate_c_string)` which calls `strlen` under the hood to determine the length, and a size check afterwards is sufficient for all non-adversarial cases.

> an intentional attacker

What intentional attackers are we trying to protect from? Do we have a requirement that the C API has to withstand non-null but invalid pointer arguments (the essence of this PR)? I am concerned about the added complexity of the still-imperfect attempts to detect them. Can it ever be perfect in the first place, and at what cost? Having to use `std::launder` is already slightly concerning, and documentation mentions needing to catch segfaults? This sounds like an arms race to me.

I believe this pointer validity requirement can better be satisfied by the high-level language APIs.

Reply:
>What intentional attackers
That is always the wrong question to ask. If you have to dream up a scenario before closing a vulnerability, you're doing it wrong.

>the essence of this PR
The practical essence of this PR is to do the equivalent of `strlen` once (in the C API where the value first appears) and to do it right.